axelor-supplychain/src/main/java/com/axelor/apps/supplychain/service/batch/BatchCheckStockComplianceWithStockRules.java

```diff
@@ -0,0 +1,204 @@
+
+import com.axelor.apps.base.AxelorException;
+import com.axelor.apps.base.db.Product;
+import com.axelor.apps.base.db.repo.BatchRepository;
+import com.axelor.apps.base.service.administration.AbstractBatch;
+import com.axelor.apps.base.service.exception.TraceBackService;
+import com.axelor.apps.stock.db.StockLocation;
+import com.axelor.apps.stock.db.StockLocationLine;
+import com.axelor.apps.stock.db.StockRules;
+import com.axelor.apps.stock.db.repo.StockRulesRepository;
+import com.axelor.apps.stock.service.StockLocationService;
+import com.axelor.apps.stock.service.StockRulesService;
+import com.axelor.apps.stock.service.app.AppStockService;
+import com.axelor.apps.supplychain.db.SupplychainBatch;
+import com.axelor.common.ObjectUtils;
+import com.axelor.db.JPA;
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import javax.persistence.TypedQuery;
+
+public class BatchCheckStockComplianceWithStockRules extends AbstractBatch {
+
+  protected final StockRulesService stockRulesService;
+  protected final AppStockService appStockService;
+  protected final StockLocationService stockLocationService;
+  protected final StockRulesRepository stockRulesRepository;
+
+  @Inject
+  public BatchCheckStockComplianceWithStockRules(
+      StockRulesService stockRulesService,
+      AppStockService appStockService,
+      StockLocationService stockLocationService,
+      StockRulesRepository stockRulesRepository) {
+    this.stockRulesService = stockRulesService;
+    this.appStockService = appStockService;
+    this.stockLocationService = stockLocationService;
+    this.stockRulesRepository = stockRulesRepository;
+  }
+
+  @Override
+  protected void process() {
+
+    int offset = 0;
+    Map<StockRules, List<StockLocationLine>> stockLocationLinesByStockRules;
+    while ((stockLocationLinesByStockRules =
+            getNonCompliantStockLocationLinesByStockRules(AbstractBatch.FETCH_LIMIT, offset))
+        != null) {
+      if (ObjectUtils.isEmpty(stockLocationLinesByStockRules)) {
+        offset += AbstractBatch.FETCH_LIMIT;
+        continue;
+      }
+      for (Map.Entry<StockRules, List<StockLocationLine>> stockLocationsByStockRule :
+          stockLocationLinesByStockRules.entrySet()) {
+        StockRules stockRule = stockLocationsByStockRule.getKey();
+        List<StockLocationLine> stockLocationLines = stockLocationsByStockRule.getValue();
+        for (StockLocationLine stockLocationLine : stockLocationLines) {
+          try {
+            processStockLocationLineNonCompliantToStockRules(stockRule, stockLocationLine);
+            incrementDone();
+          } catch (Exception e) {
+            incrementAnomaly();
+            TraceBackService.trace(e, null, batch.getId());
+          }
+        }
+        offset += AbstractBatch.FETCH_LIMIT;
```

Comment:
Please manage cache here (adding a Jpa.clear() should be enough I think ?), need to test again then with more than FETCH_LIMIT entries